File: exemplos.c
// =========================================
// Arquivo: funcoes.c
// Trabalho - Comparação com Haskell
// =========================================

#include <stdio.h>
#include <string.h>

// -----------------------------------------
// 1. Dobro de um número
// -----------------------------------------
int dobro(int x) {
return x * 2;
}


// -----------------------------------------
// 2. Verificar se um número é par
// -----------------------------------------
int ehPar(int n) {
return n % 2 == 0;
}


// -----------------------------------------
// 3. Fatorial (recursão)
// -----------------------------------------
int fatorial(int n) {
if (n == 0) return 1;
return n * fatorial(n - 1);
}


// -----------------------------------------
// 4. Soma de elementos de um array
// -----------------------------------------
int somaLista(int arr[], int tamanho) {
int soma = 0;
for (int i = 0; i < tamanho; i++) {
soma += arr[i];
}
return soma;
}


// -----------------------------------------
// 5. Filtrar nomes (simulação)
// -----------------------------------------
void filtrarNomes(char nomes[][50], int tamanho) {
for (int i = 0; i < tamanho; i++) {
if (strlen(nomes[i]) > 5) {
printf("%s\n", nomes[i]);
}
}
}


// -----------------------------------------
// 6. QuickSort
// -----------------------------------------
void quicksort(int arr[], int inicio, int fim) {
int i = inicio, j = fim;
int pivo = arr[(inicio + fim) / 2];

while (i <= j) {
while (arr[i] < pivo) i++;
while (arr[j] > pivo) j--;

if (i <= j) {
int temp = arr[i];
arr[i] = arr[j];
arr[j] = temp;
i++;
j--;
}
}

if (inicio < j) quicksort(arr, inicio, j);
if (i < fim) quicksort(arr, i, fim);
}


// -----------------------------------------
// 7. Fibonacci (sem lazy evaluation)
// -----------------------------------------
int fib(int n) {
if (n <= 1) return n;
return fib(n - 1) + fib(n - 2);
}


// -----------------------------------------
// 8. Simulação de "SELECT"
// -----------------------------------------
typedef struct {
int id;
char nome[50];
} Usuario;

void buscarId(Usuario usuarios[], int tamanho, int idBusca) {
for (int i = 0; i < tamanho; i++) {
if (usuarios[i].id == idBusca) {
printf("%s\n", usuarios[i].nome);
}
}
}


// -----------------------------------------
// 9. Função de função (simulação)
// -----------------------------------------
int somarCinco(int x) {
return x + 5;
}

int dobro2(int x) {
return x * 2;
}

int dobroMaisCinco(int x) {
return dobro2(somarCinco(x));
}


// -----------------------------------------
// 10. Currificação (simulação)
// -----------------------------------------
int soma(int x, int y) {
return x + y;
}

// "Aplicação parcial" simulada
int somar10(int y) {
return soma(10, y);
}


// -----------------------------------------
// Função principal (teste)
// -----------------------------------------
int main() {

printf("Dobro: %d\n", dobro(5));
printf("Eh par: %d\n", ehPar(4));
printf("Fatorial: %d\n", fatorial(5));

int lista[] = {1, 2, 3, 4};
printf("Soma lista: %d\n", somaLista(lista, 4));

char nomes[][50] = {"Ana", "Camila", "Joao", "Fernanda"};
printf("Nomes com mais de 5 letras:\n");
filtrarNomes(nomes, 4);

int arr[] = {5, 3, 8, 1, 2};
quicksort(arr, 0, 4);
printf("QuickSort: ");
for (int i = 0; i < 5; i++) {
printf("%d ", arr[i]);
}
printf("\n");

printf("Fibonacci (10): %d\n", fib(10));

Usuario usuarios[] = {
{1, "Guilherme"},
{2, "Ane"},
{3, "Maria"},
{4, "Camila"}
};
printf("Buscar ID 2:\n");
buscarId(usuarios, 4, 2);

printf("Dobro + cinco: %d\n", dobroMaisCinco(3));
printf("Somar 10: %d\n", somar10(5));

return 0;
}

File: exemplos.hs
-- =========================================
-- Arquivo: Funcoes.hs
-- Trabalho de Haskell - Exemplos práticos
-- =========================================

-- 1. Dobro de um número
-- Recebe um número e retorna o dobro
dobro :: Int -> Int
dobro x = x * 2


-- 2. Verificar se um número é par
-- Retorna True se for par, False caso contrário
ehPar :: Int -> Bool
ehPar n = mod n 2 == 0


-- 3. Fatorial com recursão
-- Caso base: 0! = 1
-- Caso recursivo: n! = n * (n-1)!
fatorial :: Int -> Int
fatorial 0 = 1
fatorial n = n * fatorial (n - 1)


-- 4. Soma de elementos de uma lista
-- Caso base: lista vazia = 0
-- Caso recursivo: soma cabeça + resto
somaLista :: [Int] -> Int
somaLista [] = 0
somaLista (x:xs) = x + somaLista xs


-- 5. Filtrar nomes com mais de 5 letras
-- Usa função de ordem superior (filter)
filtrarNomes :: [String] -> [String]
filtrarNomes nomes = filter (\n -> length n > 5) nomes


-- 6. QuickSort (ordenar lista)
-- Divide a lista em menores e maiores que o pivô
quicksort :: [Int] -> [Int]
quicksort [] = []
quicksort (pivo:resto) =
quicksort [x | x <- resto, x <= pivo] ++
[pivo] ++
quicksort [x | x <- resto, x > pivo]


-- 7. Fibonacci infinito (Lazy Evaluation)
-- Lista infinita calculada sob demanda
fibs :: [Int]
fibs = 0 : 1 : zipWith (+) fibs (drop 1 fibs)


-- 8. Simulação de SELECT (banco de dados)
-- Lista de usuários (id, nome)
usuarios :: [(Int, String)]
usuarios = [(1, "Guilherme"), (2, "Ane"), (3, "Maria"), (4, "Camila")]

-- Busca usuários por id
buscarId :: Int -> [(Int, String)]
buscarId idLista = filter (\(i, nome) -> i == idLista) usuarios


-- 9. Função de função (composição)
-- Combina duas funções usando (.)
somarCinco :: Int -> Int
somarCinco x = x + 5

dobroMaisCinco :: Int -> Int
dobroMaisCinco = dobro . somarCinco
-- Exemplo: dobroMaisCinco 3 = (3 + 5) * 2 = 16


-- 10. Função currificada
-- Recebe um argumento por vez
soma :: Int -> Int -> Int
soma x y = x + y

-- Aplicação parcial (fixa um argumento)
somar10 :: Int -> Int
somar10 = soma 10
-- Exemplo: somar10 5 = 15


-- =========================================
-- Função principal (para testar tudo)
-- =========================================

main :: IO ()
main = do
print (dobro 5)
print (ehPar 4)
print (fatorial 5)
print (somaLista [1,2,3,4])
print (filtrarNomes ["Ana", "Camila", "Joao", "Fernanda"])
print (quicksort [5,3,8,1,2])
print (take 10 fibs)
print (buscarId 2)
print (dobroMaisCinco 3)
print (somar10 5)